### PR TITLE
[build][qt] Add support for qmapboxgl installation

### PR DESCRIPTION
**Pull request by rinigus**
This PR allows to install Qt version of the library in the system for use in the 3rd party programs without Qt Location wrapper. For that, `make install` has been added for Qt.

In addition, to avoid non-required dependencies while building the library, Qt platform can be built without test, demo code, and without headless support. The latter was added as an option to avoid dependency on Qt Widgets and Qt OpenGL.

### CODE REVIEW

**Comment by rinigus** [comment]
Fixes #16328 for Qt

File: platform/qt/qt.cmake [hunk 1]
```diff
@@ -1,9 +1,18 @@
 # Note: Using Sqlite instead of QSqlDatabase for better compatibility.
 
+option(MBGL_WITH_QT_DEMO "Build Mapbox GL Qt demo" ON)
+option(MBGL_WITH_QT_TEST "Build Mapbox GL Qt tests" ON)
+option(MBGL_WITH_QT_HEADLESS "Build Mapbox GL Qt with headless support" ON)
```
Comment (tmpsantos) [inline platform/qt/qt.cmake:5]: Isn't there a way to call the install rule for just a single target?

Comment (rinigus) [reply platform/qt/qt.cmake:5]: Problem is that even if I just want to build qmapboxgl, I get dependencies pulled into it that I don't want to have. For example, Qt::Widgets are not used on our embedded platform (Sailfish OS). So, it was the best way I found to make it possible.

**Comment by rinigus** [comment]
Bump. Would it be possible to review this, please. I can clean up and rebase as needed

**Comment by rinigus** [comment]
Can someone still review the PR and let me know if I should resubmit it. @tmpsantos, would you have time for it? Would like to have an opportunity to install the library using `make install` - something that is missing currently and addressed here

**Comment by rinigus** [comment]
Ping: Can someone still review the PR and let me know if I should resubmit it. @tmpsantos, would you have time for it? Would like to have an opportunity to install the library using make install - something that is missing currently and addressed here

File: CMakeLists.txt [hunk 2]
```diff
@@ -1051,3 +1051,9 @@ add_subdirectory(${PROJECT_SOURCE_DIR}/render-test)
 if(EXISTS ${PROJECT_SOURCE_DIR}/internal/internal.cmake)
     include(${PROJECT_SOURCE_DIR}/internal/internal.cmake)
 endif()
+
+# installs
+include(GNUInstallDirs)
+install(DIRECTORY include/mbgl
+        DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
+        COMPONENT development)
```
Comment (tmpsantos) [inline CMakeLists.txt:1059]: This install rule will leak to other platforms that are not GNU. Let's confine it to Qt.

File: platform/qt/qt.cmake [hunk 3]
```diff
@@ -141,60 +165,85 @@ target_link_libraries(
...
+
+install(FILES
+        platform/qt/include/QMapbox
+        platform/qt/include/QMapboxGL
+        platform/qt/include/qmapbox.hpp
+        platform/qt/include/qmapboxgl.hpp
+        DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/qt5"
+        COMPONENT development)
+
+if (MBGL_WITH_QT_DEMO)
```
Comment (tmpsantos) [inline platform/qt/qt.cmake:181]: s/if (/if(

File: platform/qt/qt.cmake [hunk 4]
```diff
@@ -141,60 +165,85 @@ target_link_libraries(
...
-add_executable(
+  if (UNIX)
+    target_link_libraries(
+      mbgl-qt
+      PRIVATE
+        pthread
+        )
+  endif()
+endif()
+
```
Comment (tmpsantos) [inline platform/qt/qt.cmake:210]: Indentation should be consistent with the rest of the project.

**Review by tmpsantos — COMMENTED** [review]
Overall I find this change makes the buildsystem a lot harder to read for something that, if you are doing a debian package, could be done on a package install rule.

As for removing dependencies such as QtWidgets, I suggest we take the reverse approach. What about we have a `option(MBGL_WITH_LIB_ONLY "Build only qmapboxgl library" OFF)` and you have in the middle of the `qt.cmake` file something like... 
```
if(MBGL_WITH_LIB_ONLY)
    return()
endif()

find_package(Qt5Widgets REQUIRED)

other targets like tests... etc...
...


```

**Comment by rinigus** [comment]
Let's agree on the target first - whether to get `make install` working or skip it.

It is not about packaging for Debian, but 
- using Mapbox GL Native from source
- simplify packaging on multiple distributions. Currently that would simplify it for Sailfish, Alpine (from which postmarketOS), Fedora. If someone wants to package it for other distributions, that would be easier as well.

As a developer, I would expect `make install` to work for a library I am developing against.

As for `MBGL_WITH_LIB_ONLY`, sure, we can add that instead of `MBGL_WITH_QT_DEMO` and `MBGL_WITH_QT_TEST`. Although, probably `MBGL_WITH_QT_HEADLESS` would be needed to allow building headless version with full GL and Widgets.

As soon as we agree on the targets, I will ask around how to replace GNUInstallDirs (or whether it is supported out of GNU) and work on the rest.

**Comment by rinigus** [comment]
@tmpsantos: revised the approach and followed your suggestions:

- GNUInstallDirs are included only in Qt. It is expected to work not only on GNU platforms, but just provides a mechanism to add install dir path, see [manual](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html)

- formatting is consistent with the whole project

- there is an option to build just lib, without demo and tests, as suggested.

Please review and let me know if something needs to be adjusted.

**Comment by rinigus** [comment]
This PR has been developed further by @PureTryOut in https://github.com/mapbox/mapbox-gl-native/pull/16591

I am closing it over here, please review changes in https://github.com/mapbox/mapbox-gl-native/pull/16591 and I hope they are acceptable.

**Pull request closed without merging**